src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,555 @@
+    if (!redundant_rfs.member(rf)) {
+      // Move RFs out of counted loops when possible.
+      IdealLoopTree* lpt = get_loop(rf);
+      Node* referent = rf->referent();
+      Node* loop_exit = lpt->unique_loop_exit_or_null();
+      if (lpt->is_invariant(referent) && loop_exit != nullptr) {
+        // Switch to the outermost loop.
+        for (IdealLoopTree* outer_loop = lpt->_parent;
+             outer_loop->is_invariant(referent) && outer_loop->unique_loop_exit_or_null() != nullptr;
+             outer_loop = outer_loop->_parent) {
+          assert(is_member(outer_loop, rf), "");
+          loop_exit = outer_loop->unique_loop_exit_or_null();
+        }
+        assert(loop_exit != nullptr, "");
+        worklist.push(referent);
+        worklist.push(loop_exit);
+        redundant_rfs.push(rf);
+      }
+    }
+  }
+
+  // Populate RFs outside counted loops.
+  while (worklist.size() > 0) {
+    Node* ctrl_out = worklist.pop();
+    Node* referent = worklist.pop();
+    insert_rf(ctrl_out, referent);
+  }
+
+  // Redundancy is determined by dominance relation.
+  // Sometimes it becomes evident that an RF is redundant once it is moved out of the loop.
+  // Also, newly introduced RF can make some existing RFs redundant.
+  find_redundant_rfs(redundant_rfs);
+
+  // Eliminate redundant RFs.
+  bool progress = (redundant_rfs.size() > 0);
+  while (redundant_rfs.size() > 0) {
+    remove_rf(redundant_rfs.pop()->as_ReachabilityFence());
+  }
+
+  assert(redundant_rfs.size() == 0, "");
+  assert(!has_redundant_rfs(redundant_rfs, true /*rf_only*/), "");
+
+  return progress;
+}
+
+//======================================================================
+//---------------------------- Phase 2 ---------------------------------
+
+// Linearly traverse CFG upwards starting at n until first merge point.
+// All encountered safepoints are recorded in safepoints list.
+static void linear_traversal(Node* n, Node_Stack& worklist, VectorSet& visited, Node_List& safepoints) {
+  for (Node* ctrl = n; ctrl != nullptr; ctrl = ctrl->in(0)) {
+    assert(ctrl->is_CFG(), "");
+    if (visited.test_set(ctrl->_idx)) {
+      return;
+    } else {
+      if (ctrl->is_Region()) {
+        worklist.push(ctrl, 1);
+        return; // stop at merge points
+      } else if (ctrl->is_SafePoint() && is_interfering_sfpt_candidate(ctrl->as_SafePoint())) {
+        safepoints.push(ctrl);
+      }
+    }
+  }
+}
+
+// Enumerate all safepoints which are reachable from the RF to its referent through CFG.
+// Start at RF node and traverse CFG upwards until referent's control node is reached.
+static void enumerate_interfering_sfpts(ReachabilityFenceNode* rf, PhaseIdealLoop* phase, Node_List& safepoints) {
+  Node* referent = rf->referent();
+  Node* referent_ctrl = phase->get_ctrl(referent);
+  assert(phase->is_dominator(referent_ctrl, rf), "sanity");
+
+  VectorSet visited;
+  visited.set(referent_ctrl->_idx); // end point
+
+  Node_Stack stack(0);
```

Comment:
`ResouceMark`?

Reply:
It's hard to place a nested ResourceMark because there are dynamically reallocated data structures with different life cycles. Instead, I moved temporary data structure allocations up in the call chain and made them shared across all RF nodes.